Diverse beam search

ayushtiku5 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Implementation of diverse beam search decoding as described in the paper: https://arxiv.org/pdf/1610.02424.pdf 

diversity function reference taken from: https://github.com/ashwinkalyan/dbs

## Implementation details
Consider a T5 summarization task.
`article="Justin Timberlake and Jessica Biel, welcome to parenthood. 
The celebrity couple announced the arrival of their son, Silas Randall Timberlake, in statements to People. 
"Silas was the middle name of Timberlake's maternal grandfather Bill Bomar, who died in 2012, while Randall is the musician's own middle name, as well as his father's first," People reports. 
The couple announced the pregnancy in January, with an Instagram post. It is the first baby for both."`

Generation using normal beam search can be done as:
`model.generate(
   input_ids=input_ids,
   num_beams=2,
   num_return_sequences=2
)`

This generates:
`['the couple announced the pregnancy in January. it is the first baby for both.', 'the couple announced the pregnancy in January. it is the first baby for both of them ']`

Generation using diverse beam search can be done as:
`model.generate(
   input_ids=input_ids,
   num_beams=2,
   num_return_sequences=2,
   beam_groups=2,
   diversity_penalty=1.5
)`

This generates:
`['the couple announced the pregnancy in January. it is the first baby for both.', 'Justin Timberlake and Jessica Biel have welcomed their son, Silas Randall ']`

This means that 2 beams will be divided into 2 groups of 1 beam each, ensuring diversity between each group. NOTE: If `beam_groups=1`, then it will be same as the normal beam search as all the beams belong to the same group. Higher `diversity_penalty` will ensure more diversity between the groups of beams. When doing generation using diverse beam search, we need to ensure that `num_beams>=beam_groups` and also `num_beams` is divisible by `beam_groups`.
 
## Who can review?

@patrickvonplaten, @TevenLeScao

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

 albert, bert, XLM: @LysandreJik
 GPT2: @LysandreJik, @patrickvonplaten
 tokenizers: @mfuntowicz
 Trainer: @sgugger
 Benchmarks: @patrickvonplaten
 Model Cards: @julien-c
 examples/distillation: @VictorSanh
 nlp datasets: [different repo](https://github.com/huggingface/nlp)
 rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)
 Text Generation: @patrickvonplaten, @TevenLeScao
 Blenderbot, Bart, Marian, Pegasus: @patrickvonplaten
 T5: @patrickvonplaten
 Rag: @patrickvonplaten, @lhoestq
 EncoderDecoder: @patrickvonplaten
 Longformer, Reformer: @patrickvonplaten
 TransfoXL, XLNet: @TevenLeScao, @patrickvonplaten
 examples/seq2seq: @patil-suraj
 examples/bert-loses-patience: @JetRunner
 tensorflow: @jplu
 examples/token-classification: @stefan-it
 documentation: @sgugger
 FSTM: @stas00
 -->

ayushtiku5 commented:
@patrickvonplaten I am implementing diverse beam search. Please do suggest code design for this. I have implemented on a high level for now. 😃

patrickvonplaten commented:
> @patrickvonplaten I am implementing diverse beam search. Please do suggest code design for this.

Awesome that you work on this!

I think this looks like the right approach! However, I'd also recommend creating a new beam_scorer to be sure to not break backwards compatilibily. We can see at a later stage if we can try to merge some code together with the current beam search code :-) 

Also, can you add a link to the paper in this PR ? this would be great :-)

ayushtiku5 commented:
@patrickvonplaten please review. I have made the required changes :)

ayushtiku5 commented:
@patrickvonplaten just a gentle reminder to review the PR. Thanks!

patrickvonplaten commented:
> @patrickvonplaten just a gentle reminder to review the PR. Thanks!

Sorry, I'll review the PR this week! Also wondering how this PR relates to this one: #8840

ayushtiku5 commented:
@patrickvonplaten I think #8840 ensures that first token of every predicted sequence is different. This PR ensures diversity between group of beams at every time step of sequence generation. I think this will be more generic. Also we can change extent of diversity using `diversity_penalty` parameter.

patrickvonplaten commented on src/transformers/generation_beam_search.py:
```diff
@@ -309,6 +309,209 @@ def finalize(
+    :class:`transformers.BeamScorer` implementing diverse beam search decoding.
+
+    Adapted in part from `Facebook's XLM beam search code
+    <https://github.com/facebookresearch/XLM/blob/9e6f6814d17be4fe5b15f2e6c43eb2b2d76daeb4/src/model/transformer.py#L529>`__.
```
Could we maybe replace this link with a fitting one in `https://github.com/ashwinkalyan/dbs`?

ayushtiku5 replied:
Added the reference for DBS implementation in doc. I have kept `Facebook's XLM beam search code` reference also since most of the part is same as beam search.

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -381,6 +383,12 @@ def generate(
                 speed up decoding.
+            beam_groups (:obj:`int`, `optional`, defaults to 1):
+                Number of groups to divide :obj:`num_beams` into in order to ensure diversity among different groups of
+                beams. See `this paper <https://arxiv.org/pdf/1610.02424.pdf>`__ for more details.
```
Could we add one sentence like `"To enable diverse beam search, :obj:`beam_groups` should be set to a value larger than 1.`

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -381,6 +383,12 @@ def generate(
+                beams. See `this paper <https://arxiv.org/pdf/1610.02424.pdf>`__ for more details.
+            diversity_penalty (:obj:`float`, `optional`, defaults to 0.0):
+                This value is subtracted from a beam's score if it generates a token same as any beam from other group
+                at a particular time.
```
Can we add ```"Note that :obj:`diversity_penalty` is only relevant if ``diverse beam search`` is enabled."```

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -492,10 +500,13 @@ def generate(
+        is_sample_gen_mode = (num_beams == 1) and (beam_groups == 1) and do_sample is True
+        is_beam_gen_mode = (num_beams > 1) and (beam_groups == 1) and do_sample is False
+        is_beam_sample_gen_mode = (num_beams > 1) and (beam_groups == 1) and do_sample is True
+        is_diverse_beam_gen_mode = (num_beams > 1) and (beam_groups > 1)
```
Can we raise if `is_diverse_beam_gen_mode and do_sample is True`:

=>

```python 
if is_diverse_beam_gen_mode and do_sample is True:
  raise ValueError("Diverse beam search cannot be used in sampling mode. Make sure that `do_sample` is set to `False`."
```

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1209,6 +1257,225 @@ def beam_sample(
+            ...    AutoModelForSeq2SeqLM,
+            ...    LogitsProcessorList,
+            ...    MinLengthLogitsProcessor,
+            ...    BeamSearchScorer,
```
```suggestion
            ...    DiverseBeamSearchScorer,
```

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1209,6 +1257,225 @@ def beam_sample(
+                    ]
+                    token_frequency = torch.bincount(previous_group_tokens, minlength=vocab_size).to(device)
+                    next_token_scores[batch_idx * group_size : (batch_idx + 1) * group_size] -= (
+                        diversity_penalty * token_frequency
```
I'm pretty sure we can put the `diversity_penalty` into the logits_processor. Could we create a `DiversityPenaltyLogitsProcessor` for this?

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1209,6 +1257,225 @@ def beam_sample(
+    def diverse_beam_search(
+        self,
+        input_ids: torch.LongTensor,
+        diversity_penalty: float,
```
This should be its own `LogitsProcessor`

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1209,6 +1257,225 @@ def beam_sample(
+
+            >>> outputs = model.diverse_beam_search(input_ids, 5.5, beam_scorer, logits_processor=logits_processor, **model_kwargs)
+
+            >>> print("Generated:", tokenizer.batch_decode(outputs, skip_special_tokens=True))
```
This example generated:

```
Generated: ['& & & & & & &&&&&&'] 
```

for me -> could you check what might be going on here?

patrickvonplaten replied:
Also in general I'm not sure if translation is the best example of diversity beam search...would summarization maybe be a better example?

ayushtiku5 replied:
But this example is working for me. Getting `Generated: ['Wie alt bist du?']`
<img width="1127" alt="Screenshot 2020-12-05 at 11 36 04 AM" src="https://user-images.githubusercontent.com/40797286/101235390-4b65a100-36ee-11eb-86a4-0b3e59caa64b.png">

patrickvonplaten replied:
works for me now too! Sorry not sure what I was doing previously

patrickvonplaten reviewed (commented):
Hey @ayushtiku5,

Thanks for the very clean PR! It already looks very nice :-) 
A couple of things we need to take a look at still:

1) I think the provided example does not work as expected - can you check what's going on there? 
2) Can you also provide a 4-liner in the PR that shows how to use diverse beam search via `generate()` to see how it differs from "normal" beam search in certain cases? 
3) It would be important to stay consistent with the `LogitsProcessor` design by making a `LogitsProcessor` class for the `Hamming diversity logic`. Let's keep this for the end though since we'll have to change the general `LogitsProcessor` design a bit for this. I'll help you here!!!

It would be awesome if you could take a look at points 1) and 2)!
Let me know if you need help :-)

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1209,6 +1257,225 @@ def beam_sample(
+                next_token_scores = F.log_softmax(next_token_logits, dim=-1)  # (batch_size * group_size, vocab_size)
+                vocab_size = next_token_scores.shape[-1]
+
+                # hamming diversity: penalise using same token in current group which was used in previous groups at
```
Actually, I think the best approach here is to make a `HammingDiversityLogitsProcessor`. We'll need to adapt the `LogitsProcessor` framework a bit for that, but it will be possible. I'll help you with this. Let's do this at the end though and first see that the function works :-)

patrickvonplaten commented on src/transformers/generation_utils.py:
```diff
@@ -1209,6 +1257,225 @@ def beam_sample(
+
+                # hamming diversity: penalise using same token in current group which was used in previous groups at
+                # the same time step
+                for batch_idx in range(batch_size):
```
would be awesome to put the function into a `LogitsProcessor`

ayushtiku5 commented:
@patrickvonplaten Also I was thinking that currently I am subtracting the diversity penalty directly from the `beam_scores`. So, finally when we are doing `beam_scorer.finalize()`, the `final_beam_scores` will also include the effect of `diversity_penalty`. 

I was thinking maybe we should penalise the `beam_scores` with diversity penalty only when we are selecting top `2*group_size` beam candidates:
`next_token_scores, next_tokens = torch.topk(
                    next_token_scores, 2 * group_size, dim=1, largest=True, sorted=True
                )`

But for choosing the final beams in the end the scores shouldn't include the penalty due to diversity. What do you think?

patrickvonplaten commented:
> @patrickvonplaten Also I was thinking that currently I am subtracting the diversity penalty directly from the `beam_scores`. So, finally when we are doing `beam_scorer.finalize()`, the `final_beam_scores` will also include the effect of `diversity_penalty`.
> 
> I was thinking maybe we should penalise the `beam_scores` with diversity penalty only when we are selecting top `2*group_size` beam candidates:
> `next_token_scores, next_tokens = torch.topk( next_token_scores, 2 * group_size, dim=1, largest=True, sorted=True )`
> 
> But for choosing the final beams in the end the scores shouldn't include the penalty due to diversity. What do you think?

Hey @ayushtiku5,

That's a good point! I do think though that we should leave the `beam_scores` as there are in the end as well. My main arguments are:

1) It helps to have more diversity in the output. If we only use the diversity penalty for choosing the next beam_token, but not add it to the `_beam_scores`, the beam_scores will be very high for beams of similar tokens, which I think is what we want to prevent here. I think `beam_scores` should be penalized for every  token in the corresponding `beam_idx` that is also present in another `beam_idx` of the same `beam_group`. It's also more consistent and logical IMO: We should update the `beam_score` with the `probability` that the current beam_id was selected. 

2) It would be very ugly to implement and I'd like to avoid it...

ayushtiku5 commented:
> > @patrickvonplaten Also I was thinking that currently I am subtracting the diversity penalty directly from the `beam_scores`. So, finally when we are doing `beam_scorer.finalize()`, the `final_beam_scores` will also include the effect of `diversity_penalty`.
> > I was thinking maybe we should penalise the `beam_scores` with diversity penalty only when we are selecting top `2*group_size` beam candidates:
> > `next_token_scores, next_tokens = torch.topk( next_token_scores, 2 * group_size, dim=1, largest=True, sorted=True )`
> > But for choosing the final beams in the end the scores shouldn't include the penalty due to diversity. What do you think?
> 
> Hey @ayushtiku5,
> 
> That's a good point! I do think though that we should leave the `beam_scores` as there are in the end as well. My main arguments are:
> 
> 1. It helps to have more diversity in the output. If we only use the diversity penalty for choosing the next beam_token, but not add it to the `_beam_scores`, the beam_scores will be very high for beams of similar tokens, which I think is what we want to prevent here. I think `beam_scores` should be penalized for every  token in the corresponding `beam_idx` that is also present in another `beam_idx` of the same `beam_group`. It's also more consistent and logical IMO: We should update the `beam_score` with the `probability` that the current beam_id was selected.
> 2. It would be very ugly to implement and I'd like to avoid it...
> 
> Is that fine for you?

@patrickvonplaten yeah sure, I am fine with this.

patrickvonplaten commented:
@ayushtiku5 - hope it's ok that I fiddled quite a bit with your PR. The functionality is kept 1:1 the same (I added an integration test in the very beginning to be sure of that), but the design is changed a bit with the main goal to keep the method as general as possible. 

IMO, the PR is now good to merge :-) Could you take a final look at whether the new names and design is ok for you? 

Afterward, we can think about a nice code snippet / use case to advertise the big new feature of `transformers` :-) 
Awesome job!

patrickvonplaten commented:
@ayushtiku5 do you think the following code snippet could be a nice use case of diverse beam search?

```python
from transformers import pipeline
summarizer = pipeline("summarization", model="sshleifer/distilbart-xsum-12-6")

ARTICLE = """Part of the Broad Road was closed to traffic on Sunday at about 18:00 GMT.
The three adults and three children have been taken to Altnagelvin Hospital with non
life-threatening injuries. The Fire Service, Northern Ireland Ambulance Service
and police attended the crash. The Broad Road has since been reopened."""

# normal beam search
summarizer(ARTICLE, num_return_sequences=2)
# => [' Five people, including three children, have been taken to hospital following a two-vehicle crash in Londonderry.',
# ' Five people, including three children, have been taken to hospital after a two-vehicle crash in Londonderry.']

# diverse beam search
summarizer(ARTICLE, num_return_sequences=2, num_beam_groups=6, diversity_penalty=10.0)
# => ['Three men are in hospital after a car and a lorry crashed in Londonderry.',
# 'Six pedestrians were injured when a car and two vehicles crashed in County Antrim.']
```

patrickvonplaten reviewed (approved):
Good to merge for me. What be nice if @LysandreJik @sgugger can check as well. Also cc @patil-suraj if you're interested.

ayushtiku5 commented:
> @ayushtiku5 - hope it's ok that I fiddled quite a bit with your PR. The functionality is kept 1:1 the same (I added an integration test in the very beginning to be sure of that), but the design is slightly different with the main goal to keep the method as general as possible.
> 
> IMO, the PR is now good to merge :-) Could you take a final look at whether the new names and design is ok for you?
> 
> Afterward, we can think about a nice code snippet / use case to advertise the big new feature of `transformers` :-)
> Awesome job!

Hey @patrickvonplaten ,

Just one thing. In the `BeamScorer`'s `finalize()` method, we are directly selecting top `num_beams` beams from the `final_beam_scores`. This assumes that the beam scores in `final_beam_scores` will be sorted in decreasing order for a particular `batch_idx`. However, this will not be the case for our diverse beam search. `final_beam_scores` will be sorted for the beams inside a particular group, but not necessarily for all the beams for a particular `batch_idx`. So, I think we will have to sort the `final_beam_scores` for every `batch_idx`. I did this previously [here](https://github.com/huggingface/transformers/pull/8627/commits/14d5b6ca6e527eac2cdb9e9400d4c00f6d7add01#diff-098eb3834a12a0788445325f6795950fc5d59ec8fc8d34fef115ae5e379e18f2R292)

The rest looks good to me. Thanks for refactoring!

patrickvonplaten commented:
Hey @ayushtiku5,

sorry I forgot to mention on that. IMO we don't need to add this functionality because it doesn't matter whether the scores are sorted or not. In this line: https://github.com/huggingface/transformers/blob/72d6c9c68ba19b2e991b0d7a32989410399b33f5/src/transformers/generation_beam_search.py#L330 you can see that the `add(...)` method automatically keeps the best scores and throws out the worse scores. Since the loop goes through all scores anyways it does not matter IMO whether they are sorted or not. 

What do you think? IMO, we can revert the last commit

ayushtiku5 commented:
> Hey @ayushtiku5,
> 
> sorry I forgot to mention on why I deleted those lines. IMO we don't need to add this functionality because it doesn't matter whether the scores are sorted or not. In this line:
> 
> https://github.com/huggingface/transformers/blob/72d6c9c68ba19b2e991b0d7a32989410399b33f5/src/transformers/generation_beam_search.py#L330
> 
> you can see that the `add(...)` method automatically keeps the best scores and throws out the worse scores. Since the loop goes through all scores anyway it does not matter IMO whether they are sorted or not.
> What do you think? IMO, we can revert the last commit.

Yeah sorry! I completely missed it. Reverted the commit.

patrickvonplaten commented:
> > Hey @ayushtiku5,
> > sorry I forgot to mention on why I deleted those lines. IMO we don't need to add this functionality because it doesn't matter whether the scores are sorted or not. In this line:
> > https://github.com/huggingface/transformers/blob/72d6c9c68ba19b2e991b0d7a32989410399b33f5/src/transformers/generation_beam_search.py#L330
> > 
> > you can see that the `add(...)` method automatically keeps the best scores and throws out the worse scores. Since the loop goes through all scores anyway it does not matter IMO whether they are sorted or not.
> > What do you think? IMO, we can revert the last commit.
> 
> Yeah sorry! I completely missed it. Reverted the commit.

No worries :-) The comment wasn't the best either - I updated it. Think it's a bit clearer now.

sgugger commented on src/transformers/generation_beam_search.py:
```diff
@@ -177,6 +186,11 @@ def __init__(
 
+        if not isinstance(num_beam_groups, int) or (num_beam_groups > num_beams) or (num_beams % num_beam_groups != 0):
+            raise ValueError(
+                f"`num_beam_groups` has to be an integer smaller or equal than `num_beams` and `num_beams` has to be divisible by `num_beam_groups`, but is {num_beam_groups} with `num_beams` being {num_beams}."
```
Nit: can we split this in two lines to respect the 119 chars max length?

sgugger commented on src/transformers/generation_logits_process.py:
```diff
@@ -400,3 +410,65 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
+    <https://arxiv.org/pdf/1610.02424.pdf>`__ for more details.
+
+    Args:
+        diversity_penalty (:obj:`float`, `optional`):
+            This value is subtracted from a beam's score if it generates a token same as any beam from other group at a
+            particular time. Note that :obj:`diversity_penalty` is only effective if ``group beam search`` is enabled.
+        num_beams (:obj:`int`, `optional`):
+            Number of beams used for group beam search. See `this paper <https://arxiv.org/pdf/1610.02424.pdf>`__ for
+            more details.
+        num_beam_groups (:obj:`int`, `optional`):
```
The signature below does not show those to be optionals, so this should be removed from the docstring.

sgugger commented on src/transformers/generation_utils.py:
```diff
@@ -1208,6 +1273,213 @@ def beam_sample(
+
+        assert (
+            num_beams * batch_size == batch_beam_size
+        ), "Batch dimension of `input_ids` should be {num_beams * batch_size}, but is {batch_beam_size}."
```
```suggestion
        ), f"Batch dimension of `input_ids` should be {num_beams * batch_size}, but is {batch_beam_size}."
```

sgugger reviewed (approved):
Looks very clean to me! Just have a few nits here and there.
Thanks a lot for your PR!

patrickvonplaten commented:
@ayushtiku5 - super sorry, we messed up the previous branch yesterday. I opened a new PR with the same authorship -> so it should be good to merge :-)
